ciphers/hill_cipher.cpp

```diff
@@ -0,0 +1,411 @@
+/**
+ * @file hill_cipher.cpp
+ * @author [Krishna Vedala](https://github.com/kvedala)
+ * @brief Implementation of [Hill
+ * cipher](https://en.wikipedia.org/wiki/Hill_cipher) algorithm.
+ *
+ * Program to generate the encryption-decryption key and perform encryption and
+ * decryption of ASCII text.
+ */
+
+#include <cassert>
+#include <cmath>
+#include <ctime>
+#include <iomanip>
+#include <iostream>
+#include <string>
+#include <valarray>
+#include <vector>
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include "../numerical_methods/lu_decomposition.h"
+
+/**
+ * operator to print a matrix
+ */
+template <typename T>
+static std::ostream &operator<<(std::ostream &out, matrix<T> const &v) {
+    const int width = 15;
+    const char separator = ' ';
+
+    for (size_t row = 0; row < v.size(); row++) {
+        for (size_t col = 0; col < v[row].size(); col++)
+            out << std::left << std::setw(width) << std::setfill(separator)
+                << v[row][col];
+        out << std::endl;
```

Comment:
```suggestion
        for (size_t col = 0; col < v[row].size(); col++) {
            out << std::left << std::setw(width) << std::setfill(separator)
                << v[row][col];
        }
        out << std::endl;
```
Same in other parts of the code.

Reply:
do not worry about it for now. This is a work in progress - a draft pull.

Reply:
@Panquesito7 wrote 400+ lines of code without a single LGTM alert :)